include/mbedtls/pkcs12.h

```diff
@@ -73,27 +73,70 @@ int mbedtls_pkcs12_pbe_sha1_rc4_128( mbedtls_asn1_buf *pbe_params, int mode,
- * \param input      the input data
- * \param len        data length
- * \param output     the output buffer
+ * \param pbe_params  The ASN.1 algorithm params.
+ * \param mode        The operation mode (either #MBEDTLS_PKCS12_PBE_ENCRYPT or
+ *                    #MBEDTLS_PKCS12_PBE_DECRYPT).
+ * \param cipher_type The cipher used.
+ * \param md_type     The message digest used.
+ * \param pwd         The password to use (may be NULL if no password is used).
+ * \param pwdlen      The length of the password at \p pwd (may be 0).
+ * \param input       The buffer with the input data to process.
+ * \param len         The length of the input data buffer at \c input.
+ * \param output      The buffer to hold the output data.
  *
- * \return           0 if successful, or a MBEDTLS_ERR_XXX code
+ * \return           \c 0 on successful.
+ * \return           A negative error code on failure.
+ */
+MBEDTLS_DEPRECATED int mbedtls_pkcs12_pbe( mbedtls_asn1_buf *pbe_params,
+                                           int mode,
+                                           mbedtls_cipher_type_t cipher_type,
+                                           mbedtls_md_type_t md_type,
+                                           const unsigned char *pwd,
+                                           size_t pwdlen,
+                                           const unsigned char *input,
+                                           size_t len,
+                                           unsigned char *output );
+#undef MBEDTLS_DEPRECATED
+#endif /* !MBEDTLS_DEPRECATED_REMOVED */
+
+/**
+ * \brief            This function performs Password Based PKCS#12 (encryption
+ *                   / decryption) for cipher-based and message digest-based
+ *                   PBEs.
+ *
+ * \param pbe_params  The ASN.1 algorithm params.
+ * \param mode        The operation mode (either #MBEDTLS_PKCS12_PBE_ENCRYPT or
+ *                    #MBEDTLS_PKCS12_PBE_DECRYPT).
+ * \param cipher_type The cipher used.
+ * \param md_type     The message digest used.
+ * \param pwd         The password to use (may be NULL if no password is used).
+ * \param pwdlen      The length of the password at \p pwd (may be 0).
+ * \param input       The buffer with the input data to process.
+ * \param len         The length of the input data buffer at \c input.
+ * \param output      The buffer to hold the output data.
+ * \param olen        The number of bytes written to the output buffer
+ *                    \p output.
+ *
+ * \return           \c 0 on successful.
+ * \return           A negative error code on failure.
  */
-int mbedtls_pkcs12_pbe( mbedtls_asn1_buf *pbe_params, int mode,
-                mbedtls_cipher_type_t cipher_type, mbedtls_md_type_t md_type,
-                const unsigned char *pwd,  size_t pwdlen,
-                const unsigned char *input, size_t len,
-                unsigned char *output );
+int mbedtls_pkcs12_pbe_ext( mbedtls_asn1_buf *pbe_params, int mode,
```

Comment:
Since we are modifying API.
What is the size of the output buffer?
How do we know what it is, and don't overlfow?
I think we should add an input paramter for output buffer size as well

Reply:
@RonEld: I agree that it is not clear now from the documentation what the size of the output buffer should be.

I just wanted to clarify why we do not have an input parameter for the output buffer size and probably why we do not have an output parameter for the number of bytes written to the output buffer. This password base encryption-based methods internally rely on the regular symmetric cryptographic primitives (AES, DES, etc) to encrypt/decrypt data. So if you are going to encrypt something, then the output buffer that holds the ciphertext should be the size of the input data aligned to a multiple of the block size of the crypto algorithm. Similarly, to decrypt something, the output buffer should be at least the same size as the input, but in fact the `olen` parameter that I added does not contain the number of bytes written. Instead, when decrypting `olen` has the size of the plaintext decoded but this does not include the padding to align the text to the block size. I believe that for these reasons the output buffer length and the `olen` arent actually "needed" because they are implicit, BUT I think its convenient to have at least the `olen`.

Reply:
Hi @andresag01 Thank you for your explanation.

yes, I am aware that the output size should be at least as input size, but if I am not mistaking, we are now adding the output buffer size as input parameter as well. It is also good for checking the `olen` hasn't exceeded the buffer size.

As for your comment on whether `olen` is needed, I disagree. I think it is needed, especially to the fact that in decryption it might be smaller than the input length, because of padding bytes removal.